File: dotnet/xmsclient/Client/Handler/ChannelCloseMethodHandler.cs
using System;
using jpmorgan.mina.common;
using log4net;
using OpenAMQ.Framing;
using OpenAMQ.XMS.Client.Protocol;
using OpenAMQ.XMS.Client.State;

namespace OpenAMQ.XMS.Client.Handler
{
    public class ChannelCloseMethodHandler : IStateAwareMethodListener
    {
        private static readonly ILog _logger = LogManager.GetLogger(typeof(ChannelCloseMethodHandler));

        private static ChannelCloseMethodHandler _handler = new ChannelCloseMethodHandler();

        public static ChannelCloseMethodHandler GetInstance()
        {
            return _handler;
        }

        public void MethodReceived(AMQStateManager stateManager, AMQMethodEvent evt)
        {
             _logger.Debug("ChannelClose method received");
            ChannelCloseBody method = (ChannelCloseBody) evt.Method;

            int errorCode = method.ReplyCode;
            string reason = method.ReplyText;
            if (_logger.IsDebugEnabled)
            {
                _logger.Debug("Channel close reply code: " + errorCode + ", reason: " + reason);
            }

            AMQFrame frame = ChannelCloseOkBody.CreateAMQFrame(evt.ChannelId);
            evt.ProtocolSession.WriteFrame(frame);
            //if (errorCode != AMQConstant.REPLY_SUCCESS.getCode())
            // HACK
            if (errorCode != 200)
            {
                _logger.Debug("Channel close received with errorCode " + errorCode + ", throwing exception");
                evt.ProtocolSession.AMQConnection.ExceptionReceived(new AMQChannelClosedException(errorCode, "Error: " + reason));
            }
            evt.ProtocolSession.ChannelClosed(evt.ChannelId, errorCode, reason);
        }
    }
}

File: java/src/org/openamq/client/handler/ChannelCloseMethodHandler.java
package org.openamq.client.handler;

import org.apache.log4j.Logger;
import org.openamq.AMQChannelClosedException;
import org.openamq.AMQException;
import org.openamq.client.protocol.AMQConstant;
import org.openamq.client.protocol.AMQMethodEvent;
import org.openamq.client.state.AMQStateManager;
import org.openamq.client.state.StateAwareMethodListener;
import org.openamq.framing.AMQFrame;
import org.openamq.framing.ChannelCloseBody;
import org.openamq.framing.ChannelCloseOkBody;

public class ChannelCloseMethodHandler implements StateAwareMethodListener
{
    private static final Logger _logger = Logger.getLogger(ChannelCloseMethodHandler.class);

    private static ChannelCloseMethodHandler _handler = new ChannelCloseMethodHandler();

    public static ChannelCloseMethodHandler getInstance()
    {
        return _handler;
    }

    public void methodReceived(AMQStateManager stateManager, AMQMethodEvent evt) throws AMQException
    {
         _logger.debug("ChannelClose method received");
        ChannelCloseBody method = (ChannelCloseBody) evt.getMethod();

        int errorCode = method.replyCode;
        String reason = method.replyText;
        if (_logger.isDebugEnabled())
        {
            _logger.debug("Channel close reply code: " + errorCode + ", reason: " + reason);
        }

        AMQFrame frame = ChannelCloseOkBody.createAMQFrame(evt.getChannelId());
        evt.getProtocolSession().writeFrame(frame);
        if (errorCode != AMQConstant.REPLY_SUCCESS.getCode())
        {
            _logger.debug("Channel close received with errorCode " + errorCode + ", throwing exception");
            throw new AMQChannelClosedException(errorCode, "Error: " + reason);
        }
        evt.getProtocolSession().channelClosed(evt.getChannelId(), errorCode, reason);
    }
}
